pyhap/accessory_driver.py

```diff
@@ -859,118 +860,99 @@ def set_characteristics(self, chars_query, client_addr):
         :type chars_query: dict
         """
         # TODO: Add support for chars that do no support notifications.
-        updates = {}
-        setter_results = {}
-        setter_responses = {}
-        had_error = False
-        had_write_response = False
-        expired = False
 
-        if HAP_REPR_PID in chars_query:
-            pid = chars_query[HAP_REPR_PID]
+        queries = chars_query.get(HAP_REPR_CHARS, [])
+
+        to_notify = (query for query in queries if HAP_PERMISSION_NOTIFY in query)
+        for query in to_notify:
+            aid, iid, ev = query[HAP_REPR_AID], query[HAP_REPR_IID], query[HAP_PERMISSION_NOTIFY]
+            char_topic = get_topic(aid, iid)
+            action = "Subscribed" if ev else "Unsubscribed"
+            logger.debug(
+                "%s client %s to topic %s", action, client_addr, char_topic
+            )
+            self.async_subscribe_client_topic(
+                client_addr, char_topic, ev
+            )
+
+        query_results = {}
+        updates_by_accessories_services = {}
+        char_to_iid = {}
+
+        expired = False
+        pid = chars_query.get(HAP_REPR_PID, None)
+        if pid is not None:
             expire_time = self.prepared_writes.get(client_addr, {}).pop(pid, None)
-            if expire_time is None or time.time() > expire_time:
-                expired = True
+            expired = expire_time is None or time.time() > expire_time
 
-        for cq in chars_query[HAP_REPR_CHARS]:
-            aid, iid = cq[HAP_REPR_AID], cq[HAP_REPR_IID]
-            setter_results.setdefault(aid, {})
+        to_update = (query for query in queries if HAP_REPR_VALUE in query or expired)
+        for query in to_update:
+            aid, iid = query[HAP_REPR_AID], query[HAP_REPR_IID]
+            value = query.get(HAP_REPR_VALUE, None)
+            write_response_requested = query.get(HAP_REPR_WRITE_RESPONSE, False)
 
-            if HAP_REPR_WRITE_RESPONSE in cq:
-                setter_responses.setdefault(aid, {})
-                had_write_response = True
+            acc = self.accessory if self.accessory.aid == aid else None
```

Comment:
This code is performance sensitive. Please unpack the accessory aid into a local var outside the loop so we don't have to look it up each loop

Reply:
Cannot be done here, as we process queries in a flat loop, so there's no guarantee that next iteration has the same accessory. (As far as I understand).

Reply:
BTW. If you're looking at changes at GitHub, it has massacred the code view. I'd advice looking at the differences side by side.

Reply:
untested, but I was thinking something like this

```diff
diff --git a/pyhap/accessory_driver.py b/pyhap/accessory_driver.py
index e6fa39c..6265ec6 100644
--- a/pyhap/accessory_driver.py
+++ b/pyhap/accessory_driver.py
@@ -885,14 +885,18 @@ class AccessoryDriver:
             expire_time = self.prepared_writes.get(client_addr, {}).pop(pid, None)
             expired = expire_time is None or time.time() > expire_time
 
-        to_update = (query for query in queries if HAP_REPR_VALUE in query or expired)
-        for query in to_update:
+        primary_accessory = self.accessory
+        primary_aid = primary_accessory.aid
+
+        for query in queries:
+            if expired or HAP_REPR_VALUE not in query:
+                continue
             aid, iid = query[HAP_REPR_AID], query[HAP_REPR_IID]
             value = query.get(HAP_REPR_VALUE, None)
             write_response_requested = query.get(HAP_REPR_WRITE_RESPONSE, False)
 
-            acc = self.accessory if self.accessory.aid == aid else None
-            acc = acc or self.accessory.accessories.get(aid)
+            acc = primary_accessory if primary_aid == aid else None
+            acc = acc or primary_accessory.accessories.get(aid)
 
             char = acc.get_characteristic(aid, iid)
 
```

Reply:
need to double check the `expire` handling here

Reply:
Oh, I didn't understand you initially then. This makes sense, I need to test.

Reply:
I have a question regarding the `continue` vs `generator`. Is this more of stylistic choice of this project, or are there any performance considerations?

Reply:
It's a performance concern.  This particular function is very hot when users are using it with lighting in Home Assistant

Reply:
Sorry for that. If I knew, I would've approached this in other way initially (apart from genexps, which I assumed ran faster that continue). 

I'm unsure if this PR as a whole does not form a regression. I think it would be wise to test it against original method.